Civi/Schema/SqlEntityStorage.php

```diff
@@ -23,7 +25,65 @@ public function __construct(string $entityName) {
   }
 
   public function writeRecords(array $records): array {
-    // TODO: Implement writeRecords() method.
+    $results = [];
+    foreach ($records as $record) {
+      $results[] = $this->writeRecord($record);
+    }
+    return $results;
+  }
+
+  /**
+   * Create or update a record from supplied params.
+   *
+   * If 'id' is supplied, an existing record will be updated
+   * Otherwise a new record will be created.
+   *
+   * @param array $record
+   *
+   * @return static
+   * @throws \CRM_Core_Exception
+   */
+  public function writeRecord(array $record): object {
+    // Todo: Support composite primary keys
+    $idField = CoreUtil::getIdFieldName($this->entityName);;
+    $op = empty($record[$idField]) ? 'create' : 'edit';
+
+    // For legacy reasons, empty values would sometimes be passed around as the string 'null'.
+    // The DAO treats 'null' the same as '', and an empty string makes a lot more sense!
+    // For the sake of hooks, normalize these values.
+    $record = array_map(function ($value) {
+      return $value === 'null' ? '' : $value;
+    }, $record);
+
+    \CRM_Utils_Hook::pre($op, $this->entityName, $record[$idField] ?? NULL, $record);
+    $entityDAO = EntityRepository::getClass($this->entityName);
+    $fields = $entityDAO::getSupportedFields();
```

Comment:
We should be able to switch lines like this from DAO over to EFv2 and avoid any reliance on the DAO.